drivers/src/pmu.rs

```diff
@@ -50,19 +52,61 @@ impl PmuInfo {
         Ok(ControlFlow::Continue(()))
     }
 
-    /// Initializes the global PmuInfo structure with the information obtained from the platform SBI.
-    pub fn init() -> SbiResult<()> {
-        let mut pmu_info = PmuInfo::new();
+    fn check_delegated_counters() -> SbiResult<u64> {
+        const TIMER_COUNTER_MASK: u64 = 0x2;
+        let delegated_counters = CSR.smcounteren.get();
+
+        // TIMER can never be valid
+        if (delegated_counters & TIMER_COUNTER_MASK) != 0 {
+            return Err(SbiError::NotSupported);
+        }
```

Comment:
As per the spec this implementation is wrong also.

When HPMDE=1, smcounteren is shadow of the mcounteren register, otherwise reads return 0. TM bit will always be set.
If a vm request a write to TIME counter (shpmcounter[1] = 1) via any SBI call, it can return an SBI error though.

Any writes to TIME shpmcounter[1] should be ignored in hardware as per the spec if it is delegated.